### ISSUE TITLE
Explain and justify each decision in the schema

### ISSUE THREAD

**Issue by AshlinHarris** [TuftsCTSI/pprl#3]
I've prepared a schema based on Anonlink's tutorial and a brief reading of the API documentation. Before deployment, any design choices in this critical section needs to be documented and reviewed. Wherever possible, values should be fixed based on our study design before any testing. Afterwards, specific weights could be adjusted based on purpose-driven test cases. Some additional fine-tuning might be appropriate based on actual data sets, but ideally these would be discussed ahead of time, taking care to avoid any overfitting.

**Comment by AshlinHarris** [comment]
The clkhash docs include a [list of references](https://clkhash.readthedocs.io/en/latest/references.html). In particular, the 2011 paper by Schnell et al. describes the main algorithm underlying Anonlink and could be helpful for designing the linkage schema. I scanned the paper and made the following notes:
- I believe the anonymous linking code (ALC) generation is pseudorandom but is deterministic if the random seed is set.
- I suspect that false negatives (2 very similar records return a linkage coefficient of 0) might be impossible for any reasonable schema definition, which would be an excellent property.
  - If false-positive/false-negative rates don't need to be balanced out, the primary concern becomes security/linkage quality (false positives).
- Security is improved by increasing number of hash functions (weight, I think) relative to the length of the Bloom filter (can I take this to mean the length of the Anonlink CLK?)
- Weights should be based on the relevance of the feature, but I'm left wondering if the length/information density of the input should be considered, either for security or for linkage quality. Would it be reasonable to take a categorical value with 2 different values and hash it 100 times into an ALC with 1000 total hashes?
- For Anonlink in particular, do I need to adjust `bitsPerFeature` based on the overall length `l` or `keysize`? This isn't evident from the example schemas, but it would be nice to see how this is handled. Is there ever a chance that not all the ngrams in a field get hashed at least once? Do some get hashed one more time based on a remainder?
  - Aside: Using the full name field ("LastName, FirstName") without stripping the comma (assuming the format is consistent) could add some weight to name order in the case of term transpositions.


It would be good to have some definitive answers to the preceding questions, either from a literature search or Anonlink code review. For the time being, it may be sufficient to use our best judgment based on the Anonlink API and build tests to help address any of the assumptions above.

**Comment by AshlinHarris** [comment]
The Anonlink `DoubleFilter` seems to refer to the Kirsch-Mitzenmacher Bloom filter optimizations[^1] used by Schnell et al. I'm not sure if it provides any benefit outside the hashing step (in our case, linking is the bottleneck). `BlakeHash` appears to refer to BLAKE2.

[^1]: Adam Kirsch and Michael Mitzenmacher. Less hashing same performance: building a better bloom filter. In Yossi Azar and Thomas Erlebach, editors, Algorithms-ESA 2006. Proceedings of the 14th Annual European Symposium: 11-13 September 2006; Zürich, Switzerland, pages 456–467, Berlin, 2006. Springer
